designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,840 @@
+  latent_data& latent_data__ = static_cast<latent_data&>(ldb);
+  ... code generated to populate latent_data__ ...
+}
+```
+
+Because it does not use anything specific to the specific model class,
+this can be implemented inthe `latent_data_base_crtp` class using the
+CRTP.
+
+```
+template <class Derived>
+class latent_data_base_crtp<Derived> {
+template <typename RNG> inline void
+set_latent_data(
+    const VectorXd& params_r, 
+    latent_data_base& ldb, 
+    RNG& base_rng, 
+    std::ostream* pstream=nullptr) const {
+	static_cast<Derived&>(ldb).set_latent_data(params_r, ldb,
+                                               base_rng, pstream);
+};
+```
+
+The concrete `latent_data` class implemented along with the model
+class will be responsible for holding the memory for any data needed
+in latent data, such as `Eigen::Matrix` or `std::vector`
+instances. By reusing the same `latent_data` variable (per thread),
+variables can be reset rather than allocating and freeing each
+evaluation.
+
+### The `latent_data_base` class
+
+The `latent_data_base` class will only know about how to 
+write out variables and variable names.
+
+```cpp
+class latent_data_base {
+  int iteration_number_;
+  int get_iteration_number() {
+    return iteration_number_;
+  }
+  void set_iteration_number(int n) {
+    iteration_number_ = n;
+	}
+  virtual void num_variables() const = 0;
+  virtual void write_variable_names(std::vector<string>& names) const = 0;
+  virtual void write_variable_values(std::vector<double>& values) const = 0;
+};
+```
+
+The CRTP class will impelement the three virtual functions using the
+CRTP in the usual way.  
+This uses inheritance from the base class, but the cost is relatively
+low in that it is only done once per iteration.  This follows the way
+Stan's `model_base` class is designed.
+
+
+### Generalized `log_prob` method
+
+The existing `log_prob` method must be updated to maintain backward
+compatibility: 
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_
+log_prob(Eigen::Matrix<T_,-1,1>& params_r, 
+         const latent_data& latent_data__,
+         std::ostream* pstream = nullptr) const;
+```
+
+In this case, the `log_prob` code generation must be updated to copy
+all of the generated data into scope.
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_ 
+log_prob(
+    Eigen::Matrix<T_,-1,1>& params_r,  
+    const latent_data_base& ldb,
+    std::ostream* pstream = nullptr) const {
+  const latent_data& latent_data = static_cast<latent_data&>(ldb);
+  double sens = latent_data__.sens;
+  double spec = latent_data__.spec;
+  ... code generate as before with sens/spec in scope ...
+}
+```
+
+Note that it is implemented generically using `latent_data_base` for
+use from outside interfaces.
+
+
+### Generalized `write_array` method
+
+The existing `write_array` method must be similarly updated, because
+now it is going to need to write the generated data.
```

Comment:
This is what I prefer to the above suggestion of a `write_variable_values` function on a latent-data-specific base class